### PR TITLE
reset cache when storage possibly change (fix init of tests).

### PR DESCRIPTION
**Pull request by cheme**
Very small pr to reset root cache when building state (in some other branch my tests did fail due to the root caching when using test externalities).

### CODE REVIEW

**Comment by bkchr** [comment]
Where is this function being used?

**Comment by cheme** [comment]
On 'assimilate' (did hit that when adding content to test_externalities in another branch), via 'apply_transaction'.

**Comment by bkchr** [comment]
So we should reset the cache there?

**Comment by cheme** [comment]
Yes, modifying the storage with the mutable access means you are changing the backend (in normal condition this never happen, storage of trie backend stay the same during the whole processing, but in tests this function can be use).
We could clear cache after changing the backend in the calling function, but it looks safer to do it here (mutable access to trie backend so potentially something change and get invalidated).

**Comment by bkchr** [comment]
Hmm all of this feels very dirty.

**Comment by cheme** [comment]
I can move the the clear_cache after calling `backend_storage_mut` but it will not prevent future issue.
I would prefer renaming the function to `backend_storage_test_modify` or something that let the user know storage from a trie backend should not be change this way.

**Comment by cheme** [comment]
I could also just remove `backend_storage_mut`, and replace it by trie backend rebuilding (thus using a new empty cache).

**Comment by cheme** [comment]
> Hmm all of this feels very dirty.

I feel that it is more `backend_storage_mut` that is a bit out of place, and if you audit usage it is mostly for building test or in memory backend as a convenience function (trie backend purpose is to implement backend that is a read api (with some inner mutable access for caching)).

**Comment by cheme** [comment]
I use this code in https://github.com/paritytech/substrate/pull/8931 , but with recent refactoring it may not be needed anymore, but in any case it is still a footgun.

**Comment by bkchr** [comment]
> but in any case it is still a footgun.

If we can remove it, perfect.

File: primitives/state-machine/src/trie_backend_essence.rs [hunk 1]
```diff
@@ -94,6 +94,9 @@ where
 
 	/// Get backend storage reference.
 	pub fn backend_storage_mut(&mut self) -> &mut S {
```
Comment (bkchr) [inline primitives/state-machine/src/trie_backend_essence.rs:96]: Do we still need this?

Comment (cheme) [reply primitives/state-machine/src/trie_backend_essence.rs:96]: :+1: I did forget it , thanks.

**Review by bkchr — APPROVED** [review]

**Comment by bkchr** [comment]
bot merge

**Pull request merged**